### DISCUSSION TITLE
default_factory argument not working correctly on optional Body and Query fields

### DISCUSSION THREAD

**Discussion by sehraramiz** [fastapi/fastapi#6123] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional

from fastapi import FastAPI, Body, Query


app = FastAPI()


def body_default():
    return -1


def query_default():
    return -2


@app.post("/")
async def root(
    my_query: Optional[int] = Query(default_factory=query_default),
    my_body: Optional[int] = Body(embed=True, default_factory=body_default),
):
    return {"body": my_body, "query": my_query}


@app.post("/error")
async def error(
    my_body: Optional[int] = Body(embed=True, default=0, default_factory=body_default),
):
    return {}
```

### Description

i want to use default_factory on Body and Query

make a post request with no body and no query parameter.
~~~curl
curl http://127.0.0.1:8000/ -X POST
~~~

1- i expect to get the default values that default_factory created 
but it returns a field required error.
2- after fixing the above issue and marking the field as nonrequired i expect to get the value of default_factory() but i get None.

3- with the second example function (/error) i think the expected behaviour of a pydantic FieldInfo is to validate the arguments and if the field has a default and also a  default_factory it should raise a runtime error.

### Operating System

Linux

### FastAPI Version

0.85.0

### Python Version

3.10.5

### Additional Context

for the "field required" error i think the problem is with [this line of dependencies/utils.py](https://github.com/tiangolo/fastapi/blob/c6aa28bea2f751a91078bd8d845133ff83f352bf/fastapi/dependencies/utils.py#L383)  it should also check if a default_factory is provided and mark the field as nonrequired.

by fixing the first issue you can proceed to call the api and get the default value. but the default value is None.
i managed to fix this by using pydantic ModelField's get_default method instead of using "default" property in [request_params_to_args](https://github.com/tiangolo/fastapi/blob/c6aa28bea2f751a91078bd8d845133ff83f352bf/fastapi/dependencies/utils.py#L610) and [request_body_to_args](https://github.com/tiangolo/fastapi/blob/c6aa28bea2f751a91078bd8d845133ff83f352bf/fastapi/dependencies/utils.py#L669) functions. 

this changes produce a problem in schema loading
~~~curl
curl http://127.0.0.1:8000/docs
~~~
![openapi](https://user-images.githubusercontent.com/14166324/191815017-501aac4d-2529-4a7f-b7a4-487af50040e7.png)

**Comment by JarroVGIT** [comment, marked as answer]
Support of `default_factory` will be unlikely anytime soon I'd say, as it is still a beta keyword in Pydantic itself:

> The default_factory argument is in beta, it has been added to pydantic in v1.5 on a provisional basis. It may change significantly in future releases and its signature or behaviour will not be concrete until v2. Feedback from the community while it's still provisional would be extremely useful; either comment on [#866](https://github.com/pydantic/pydantic/issues/866) or create a new issue.

There is quite some logic going on in FastAPI, using `DefaultPlaceholder` and such to determine if a field was set or not, so I would imagine it might not be trivial to implement it yourself (as you have noticed).

**Comment by tiangolo** [comment]
Thanks for the help here @JarroVGIT ! 👏  🙇 

Thanks for reporting back and closing the issue @sehraramiz  👍

**Comment by Booplicate** [comment]
Since fastapi is now using pydantic v2, does it make sense to add support for this?
